Hotfix fix simple optimizer cull below

cliff-bohm opened the pull request:
updated algorithm for cullRemap function. Better config messages.

cgnitash commented on Optimizer/SimpleOptimizer/SimpleOptimizer.cpp:
```diff
@@ -151,37 +152,40 @@ void SimpleOptimizer::optimize(std::vector<std::shared_ptr<Organism>> &populatio
 	deltaScore = maxScore - cullBelowScore;
-	std::cout << "  cullBelowScore: " << cullBelowScore << "  deltaScore: " << deltaScore << std::endl;
+	//std::cout << "  cullBelowScore: " << cullBelowScore << "  deltaScore: " << deltaScore << std::endl;
 	for (int i = 0; i < (int)population.size(); i++) {
```
We need to replace all instances of c-style casts, with the appropriate c++ cast; which in most cases is simply a static_cast. I have this in my list of things to change in rewrite, but I haven't gotten to all of them yet (precisely because one of the reasons c-style casts are bad, is that they are painful to search for). PRs are definitely a good place to start. '
I'm aware this seems pedantic, but just google "c++ c-style casts", and you'll see what I mean.

JorySchossau replied:
Googled (from a SO post) because I couldn't remember the reasons either:
"
- C++ style casts are checked by the compiler. C style casts aren't and can fail at runtime
- also, c++ style casts can be searched for easily, whereas it's really hard to search for c style casts
- Another big benefit is that the 4 different C++ style casts express the intent of the programmer more clearly. "

cgnitash commented:
Line 180 is definitely fine. && has strictly lower precedence than !=.

cliff-bohm commented:
converted type casts to c++ style casts (or changed to size_t if it made sense).
added bool check variables to remove min != max double comparisons (which are unsafe)
added some parentheses to make Jory feel less nervous.

cliff-bohm commented:
I'm not seeing the == issues on lines 161 and 187:

I'm seeing:

line 161:
if (cullBelow >= 0 && scoresHaveDelta){ // cull and normalize scores if min == max then all scores are the same, do nothing!

line 187:
if ((cullRemap != -1) && (culledScoresHaveDetla)) {

you may need to refresh a window...

cgnitash commented:
(cullBelow >= 0) is the same as (cullBelow > 0 || cullBelow >= 0)
cullBelow is a double, you shouldn't use it for equality comparisons.

Ditto with  (cullRemap != -1)

cliff-bohm commented:
I see that, but the value will be -1 or [0,1]
I guess in this case I will use >.5 and < .5

Does this, in general, mean that we should not be using a value like this.
i.e. currently, -1 = off, [0,1] means on and value is assigned value.

The alternative is to have different  parameters for use and values:
cullBelowSwitch
cullBelowValue
cullRemapSwitch
cullRemapValue

but this seems like a lot of unneeded parameters.

I understand the point, but since this is being assigned directly, and not by a math operation, is it really not safe?

cliff-bohm commented:
converted lines to comparisons to -.5

cgnitash commented:
We can totally have a value like "-1 = off, [0,1] means on and value is assigned value". Having a bunch of extra Parameters is not a good solution.
In this case, simply "<0 means off" will suffice. It's really only the == that's a problem.

Yes... if the value is only ever set once, then it shouldn't be a problem. But as it stands, there's no way to enforce that the value will never change. If the Parameter is localised into a const variable, then there's some enforcement. But again, if the const is removed, everything will compile just fine, until it breaks completely unexpectedly. One way around this would be if there was a concept of a const ParameterLink, but since we don't I'd rather just not do == comparisons.

cgnitash commented:
Remove the question tag, if you don't have questions and then we can merge.
